### PR TITLE
Added solutions for chapter 1, exercise 1 to 5 and chapter 2 exercise 1 to 3

### PR DESCRIPTION
**Pull request by Mironor**
Given #5 here are some solutions to the exercises solved so far (for ch1 and some of ch2) just to show that there is some progress in the 'exercise' package.
I will submit solutions to following chapters as soon as they are solved.

### CODE REVIEW

**Comment by Mironor** [comment]
Actually, blocked by this line (ch. 2 ex. 4):

>Then, implement a producer thread that transfers a range of numbers 0 until 15 to the consumer thread that prints them.

It looks like it's something simple, but I only found Producer/Consumer implementations with Akka so far. Can you give some hint to the problem, please?

File: src/main/scala/org/learningconcurrency/exercises/ch1.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package org.learningconcurrency.exercises
+
+object CH1Solutions extends App{
+
+  def compose[A, B, C](g: B => C, f: A => B): A => C = g compose f
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/ch1.scala]: Nice.
We should also add the alternative solution `g(f(x))`.

File: src/main/scala/org/learningconcurrency/exercises/ch1.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  def fuse[A, B](a: Option[A], b: Option[B]): Option[(A, B)] = for {
+    aVal <- a
+    bVal <- b
+  } yield (aVal, bVal)
+
+  def check2[T](xs: Seq[T])(pred: T => Boolean): Boolean = xs.forall {x => 
+    try {
+      pred(x)
+    } catch { 
+      case _:Exception => false
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/ch1.scala]: Space before the type - http://docs.scala-lang.org/style/types.html.

File: src/main/scala/org/learningconcurrency/exercises/ch1.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    }
+  )
+
+  def permutations(x: String): Seq[String] = x.permutations.toList
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/ch1.scala]: This is a correct solution.

However, we would like to encourage the readers to use recursive functions along with for-comprehensions here.

File: src/main/scala/org/learningconcurrency/exercises/ch2.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      log(bVal.toString())
+    }
+
+    t1.join()
+    t2.join()
+
+    (aVal, bVal)
+  }
+
+  def periodically(duration: Long)(f: () => Unit): Unit = {
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/ch2.scala]: The method `periodically` must start a thread that executes a computation block every `duration` milliseconds, but the method itself should not block. It should immediately return. Also, the thread that gets started should invoked `f` indefinitely, not just `3` times.

File: src/main/scala/org/learningconcurrency/exercises/package.scala [hunk 5]
```diff
@@ -1,10 +1,14 @@
 package org.learningconcurrency
 
 
-
-
-
-
 package object exercises {
 
+  def thread(body: =>Unit): Thread = {
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/package.scala]: You may instead reuse the `thread` construct from the `ch2` package, by importing it in your exercise solutions.

**Comment by axel22** [comment]
I recommend that we keep the exercises for each chapter in separate packages, just like the chapter examples:

    package org.learningconcurrency
    package exercises
    package ch1

The exercises themselves should be in separate files, for example `Ex1.scala`, `Ex2.scala`, and so on. This will allow easily finding them, and will unclutter the code, as `Ex1.scala` can contain multiple solutions for that exercise.

**Comment by axel22** [comment]
(I can also do the refactoring to separate files for you, if you want)

**Comment by axel22** [comment]
>   > Then, implement a producer thread that transfers a range of numbers 0 until 15 to the consumer thread that prints them.

> It looks like it's something simple, but I only found Producer/Consumer implementations with Akka so far. Can you give some hint to the problem, please?

Sure! The producer/consumer pattern is not specific to Akka and the actor model. It is a more general pattern in which one computation produces some values, and another computation takes those values and does something with them. Here, the consumer depends on the values created by the producer, and cannot proceed until they become available.

In this exercise, you need to:
- create a SyncVar
- create a thread that repetitively `put`s values into the SyncVar after getting `true` from `isEmpty`
- create a thread that repetitively does the opposite

Which one is the consumer, and which is the producer here?

**Comment by Mironor** [comment]
Thanks for taking time reviewing the solutions!

I just updated the PR with regards to your comments. If you could refactor the code with separate files, it would be great, this way I could see what exact file structure you have in mind.

**Comment by Mironor** [comment]
Just saw the hint on the exercise 4, that's exactly what I was looking for, thanks!

**Comment by axel22** [comment]
You're welcome, and thank you for contributing! :)

Here's the structure I meant:

https://github.com/concurrent-programming-in-scala/learning-examples/commit/c0a4981c7f8351ba9e6e6ac31c12fc1779fcf4a6

**Pull request merged**